### PR TITLE
Danielle M Hotel

### PR DESCRIPTION
**Pull request by danimetz**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I needed to decide what classes I wanted to have. What type of methods each class was going to have and if there were going to be any dependencies on another class. I was considering having having a calendar class that organized all the dates. and new what room's were reserved and when. I thought it made the most sense to use rooms and reservations because reservations would know about blocks and rooms would just know when they were taken. 
Describe a concept that you gained more clarity on as you worked on this assignment. | I learned more about how design can really change how your code will function. I got better clarity on how dependencies make it harder to change your code because you'll need to change it in multiple places. 
Describe a _nominal test_ that you wrote for this assignment. |  Cannot book a reservation in the same room if there is another reservation on that is already existing
Describe an _edge case test_ that you wrote for this assignment. | In reservation manager I needed to test that if they put dates that matched the checkout dates of another that room would be pick the same room.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I felt like I had a pretty could idea of what each method i wanted to create in the different classes. I found it hard to write tests  before every method sometimes. I liked to use guard when I was writing tests because it would automatically run and show me that I wrote a test that failed and then I would go into my code and fill in the method that would complete it. It was cool to see it turn green after I was able to get method to run correctly. I had a hard time feeling like I wrote enough tests. Did I think of all the possible edge cases? Sometimes it was hard for me to make sure i knew my code was doing everything i wanted it to without any output. There are still a lot of things I would like to change but it performing how I would like.

### CODE REVIEW

File: lib/room.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  end
+
+  def self.list
+    return ObjectSpace.each_object(self).to_a
+  end
+
+  def is_available?(check_in,check_out)
+    new_res_dates = (check_in...check_out).to_a
+    unavailable_dates.each do |reservation|
+      reservation = (reservation[:check_in]...reservation[:check_out]).to_a
```
Comment (CheezItMan) [inline lib/room.rb:28]: Couple of things:
1.  This definitely works and lets you tell if one of the existing reservations conflicts with the given checkin checkout dates.
2.  It's a _little_ inefficient.  You have to generate a bunch of arrays and then check to see if that array contains a given date, which requires another loop.  This could be done with some comparisons to the start and end dates instead.  

BTW #2 is a minor criticism

File: lib/room.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    end
+    return true
+    # if date_range.all? { |date| unavailable_dates.include?(date) }
+    #   return 0
+    # else
+    #   return self.id
+    # end
+  end
+
+  def add_unavailablity(check_in,check_out)
```
Comment (CheezItMan) [inline lib/room.rb:36]: I would suggest adding a check to make sure the reservation is available before you add it in.  Otherwise someone using this class could get a room into an invalid state, where two reservations occur on the same dates.

File: refactors.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  RESERVATION CLASS
+    -date? method... what is the date? It should be a little bit more specific about what its checking and in this case its returning true if a reservation includes the date
+    -Couldn't decided if there needed to be a new class for the block reservations. But I allowed the availability to do blocks within my reservation class
+    -Look into the methods are they doing to much? I like that my make_reservation method will also handle if someone were trying to reserve the block room, but maybe that's doing too much?
+    -The make reservation method will always make a reservation in the first room available for a given date range, but lets say someone wants a particular room, or what if a block of rooms wants to be next to each other if those rooms are available? Maybe adding functionality that would offer solutions to these in the reservation manager.
+    -Take in a hash instead of all the arguments
+  RESERVATION MANAGER CLASS
+    -Maybe it should just be called Reservation_Manager instead of MGr
```
Comment (CheezItMan) [inline refactors.txt:12]: 👍

File: lib/reservation_mgr.rb [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  end
+
+  def available_rooms(check_in,check_out,rooms: 1)
+    if check_out < check_in
+      raise StandardError.new('The check-out date is before the check-in')
+    end
+
+    available_rooms = []
+    @rooms.each do |room|
```
Comment (CheezItMan) [inline lib/reservation_mgr.rb]: You could also do `@rooms.select` to do the same thing here.

File: lib/reservation_mgr.rb [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    else
+      block_rooms = find_available_block_rooms(block_id)
+
+      if check_in != block_rooms[0].check_in || check_out != block_rooms[0].check_out
+        raise ArgumentError.new('The dates do not match for this given block reservation')
+
+      end
+
+      rooms.times do |index|
```
Comment (CheezItMan) [inline lib/reservation_mgr.rb:42]: You should also check to see if `block_rooms` has at least `room` number of rooms available in the block.

File: lib/reservation_mgr.rb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+      end
+
+      return block_rooms[0...rooms]
+    end
+
+  end
+
+  def reservations_by_date(date)
+    reservations_on_date = []
+    @reservations.each do |reservation|
```
Comment (CheezItMan) [inline lib/reservation_mgr.rb]: Another good use for `.select`

File: spec/reservation_mgr_spec.rb [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+      expect{@new_ResMgr.make_reservation(Date.parse("2018-09-05"),Date.parse("2018-09-07"),rooms: 2)}.must_raise ArgumentError
+    end
+    it "will not reserve a room on the same day if the dates are inside the date range of a previous reservation" do
+      @new_ResMgr.make_reservation(Date.parse("2018-09-04"),Date.parse("2018-09-08"))
+      @new_ResMgr.make_reservation(Date.parse("2018-09-05"),Date.parse("2018-09-07"))
+      expect(@new_ResMgr.reservations.length).must_equal 2
+      expect(@new_ResMgr.rooms[0].unavailable_dates.length).must_equal 1
+      expect(@new_ResMgr.rooms[1].unavailable_dates.length).must_equal 1
+    end
+    it "will add two instances of res to one room if they are not on the same date" do
```
Comment (CheezItMan) [inline spec/reservation_mgr_spec.rb:50]: You are also making a reservation that begins on the date the previous one ended.  I would be specific that you are testing that feature.

File: spec/reservation_mgr_spec.rb [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+      block = @new_ResMgr.reserve_block(Date.parse("2018-09-05"),Date.parse("2018-09-07"),5,"Metzner")
+      expect(block.length).must_equal 5
+      expect(@new_ResMgr.reservations[0].block_available).must_equal true
+      @new_ResMgr.make_reservation(Date.parse("2018-09-05"),Date.parse("2018-09-07"))
+      expect(@new_ResMgr.reservations.last.block_available).must_equal false
+    end
+    it "will raise an argument error if they try to block more than 5 rooms" do
+      expect{@new_ResMgr.reserve_block(Date.parse("2018-09-05"),Date.parse("2018-09-07"),6,"Metzner")}.must_raise ArgumentError
+    end
+    it "will raise an arguemnt error if there are not enough available rooms for a given date range" do
```
Comment (CheezItMan) [inline spec/reservation_mgr_spec.rb:134]: arguemnt => argument

File: spec/reservation_mgr_spec.rb [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+  end
+
+  describe "find_available_block_rooms method" do
+    before do
+      @new_ResMgr = ReservationMgr.new(6)
+    end
+    it "returns rooms in block that have not been reserved" do
+      block = @new_ResMgr.reserve_block(Date.parse("2018-09-05"),Date.parse("2018-09-07"),5,"Metzner")
+      expect(@new_ResMgr.find_available_block_rooms("Metzner").length).must_equal 5
+      expect(@new_ResMgr.find_available_block_rooms("Metzner")[0]).must_be_kind_of Reservation
```
Comment (CheezItMan) [inline spec/reservation_mgr_spec.rb:149]: I would also check that the available block rooms doesn't include the one just reserved.

File: spec/reservation_spec.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  end
+  describe "Reservation.date method" do
+    it "Returns true if reservation includes given date" do
+      new_Res = Reservation.new("2018-09-05","2018-09-07",1)
+      expect(new_Res.date?("2018-09-05")).must_equal true
+      expect(new_Res.date?("2018-09-06")).must_equal true
+      expect(new_Res.date?("2018-09-07")).must_equal true
+
+    end
+    it "Returns false if there are no instances of reservations on that day" do
```
Comment (CheezItMan) [inline spec/reservation_spec.rb:51]: You should also check if a reservation ends on that date (should return `false`), not true like above

File: lib/room.rb [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+  def self.create(number_rooms)
+    rooms = []
+    number_rooms.times do |room_number|
+      rooms << Room.new(room_number + 1)
+    end
+    return rooms
+  end
+
+  def self.list
```
Comment (CheezItMan) [inline lib/room.rb:21]: Since you're not using this method anywhere else in your program, you can ditch it.

File: spec/reservation_mgr_spec.rb [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    end
+    it "will reserve a block room that is available if someone puts in a block party" do
+      @new_ResMgr.reserve_block(Date.parse("2018-09-05"),Date.parse("2018-09-07"),2,"Metzner")
+      res = @new_ResMgr.make_reservation(Date.parse("2018-09-05"),Date.parse("2018-09-07"),block_id: "Metzner")
+      expect(res[0].block_available).must_equal false
+    end
+    it "will raise an argument error if the block room checkin checkout dates don't match what they want to reserve" do
+      @new_ResMgr.reserve_block(Date.parse("2018-09-05"),Date.parse("2018-09-07"),2,"Metzner")
+      expect{@new_ResMgr.make_reservation(Date.parse("2018-09-06"),Date.parse("2018-09-07"),block_id: "Metzner")}.must_raise ArgumentError
+    end
```
Comment (CheezItMan) [inline spec/reservation_mgr_spec.rb:68]: You should also test what happens when you try to reserve more rooms in the block than are in that block.

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| Good number of commits and good commit messages
Answer comprehension questions	| Check
**Design** |
Each class is responsible for a single piece of the program | Well done, although your `Room` class is also tracking reservations alongside your `ReservationMgr` class.
Classes are loosely coupled | `ReservationMgr` is the only tightly coupled class.
**Wave 1** |
List rooms | Check
Reserve a room for a given date range | Check
List reservations for a given date | Check
Calculate reservation price | Check
Invalid date range produces an error | Check
Test coverage | 99%
**Wave 2** |
View available rooms for a given date range | Check
Reserving a room that is not available produces an error | Check
**Wave 3** |  Interesting how you made blocks a field in reservation, it did require `ReservationMgr` to do a bit more work.
Create a block of rooms | Check
Check if a block has rooms | Check
Reserve a room from a block | Check
**Fundamentals** | 
Names variables, classes and modules appropriately | Check
Understanding of variable scope - local vs instance | Check
Can create complex logical structures utilizing variables | Check
Appropriately uses methods to break down tasks into smaller simpler tasks | Check
Understands the differences between class and instance methods | 
Appropriately uses iterators and enumerables | You use `.each` a lot where you could simplify your code with Enumerables like `.select` or `.reject`
Appropriately writes and utilizes classes | Check
Appropriately utilizes modules as a mechanism to encapsulate functionality | **No Modules Used**
**Wrap Up** |
There is a refactors.txt file | Check
The file provides a roadmap to future changes | Check
**Additional Feedback** | Awesome work.  You implemented all the functionality with an original and effective design.  You could break down some of your methods a bit more, and make better use of Enumerables, but you had a good design and clear, readable code.  You also had effective tests with some small suggestions in my inline comments.  Well done, you hit all the learning goals for the project.  Excellent work!

File: lib/reservation_mgr.rb (hunk 4 above)
Comment (danimetz) [reply lib/reservation_mgr.rb]: Thanks! I didn't even think about using the is_available? functionality with the select enumerable!
